config/helpers.php

```diff
@@ -173,7 +169,6 @@ function esc(string $string, string $context = 'html'): string
 	 *
 	 * @param mixed $key The key to look for. Pass false or null to return the entire request array.
 	 * @param mixed $default Optional default value, which should be returned if no element has been found
-	 * @return mixed
 	 */
 	function get($key = null, $default = null)
```

Comment:
This could be type-hinted with `: mixed` as well for clarity.

Reply:
I know but https://github.com/getkirby/kirby/pull/4959#discussion_r1056177720 🙈

Reply:
Ah, right. I forgot about that conversation.

TBH I feel like `mixed` *does* add value because the user knows that the method may actually return any type (in contrast to a method where we forgot to document the return type or still need to clean the type hints).

Pinging @distantnative for thoughts.

Reply:
There's no harm in not adding it but I'm definitely in favor of adding it. I think specifying the mixed type rather than nothing is very useful both for reading the code and for the IDE.

Reply:
I think for helpers it's ok, but what you need to keep in mind is that adding a return type hint can be quite the breaking change. If we add it on a class method and someone extended the class but their method now doesn't have the same signature, it'll break completely.

So we said we would only want to do this for places where adding a type hint really has a benefit, more than just a reminder for us. I still think not adding mixed has more benefits than adding it.

Reply:
If we really don't want to do this for methods, I think we shouldn't do it for helpers either.

I guess we just disagree on whether we want to do this in general. We also add type hints for other return types all the time in the cleanup PRs. Is `mixed` different here?

Reply:
Yes, other type hints actually help us to make methods safer (as they check that only those types are returned), better UX in IDEs etc.

Adding `mixed` doesn't add anything functionally, just visually.

Reply:
We also arrived at this decision because I had added `mixed` in various places and got feedback during an RC that now many extended classes of developers broke. For too little benefit, in
my view.

Reply:
An article about that 😅 
https://devalmonte.com/blog/2018-10-10/why-we-need-mixed-types-in-php/

Reply:
Even after reading the article I don't see how this little added explicitness compensates the breaking character of such changes.

Reply:
The `mixed` type *does* show up in IDEs and it also makes it clearer for us that the type really is mixed. Alternatively we could also keep the docblock type hints for that, but TBH I don't think that's the right way considering that PHP natively supports it.

We had those breaking changes in that one PR that we reverted, but that was a special case IMO. 95% of classes are not extended from plugins. And even if they are, `mixed` hints are not going to be the only problems. As I wrote we are already in the process of adding all the other types anyway.

Reply:
Yes, my point was that the same shows up in IDEs also when you don't add it. Not specified = mixed, if anything that isn't mixed is specified.

But if you all don't mind more potential for breaking changes, let's do it. I am not against `mixed` - I just was afraid to break too much for much less benefit than e.g. type hinting a string.

Reply:
If `mixed` was added to all the clean up PRs we are doing now, what kind of breaks would happen. There is no difference between adding `string` or `mixed` to the return type of a method in a non-extended class. On the contrary, if it is an extended class, whatever you add to the return type will be broken. I don't think this is a situation related to `mixed`.

Reply:
Yes, but we are turning in circles. I keep saying that `string` etc. also are breaking changes but they add a significant benefit of type checking, while adding `mixed` functionally is the same as adding no hint (also IDEs would already assume mixed if nothing is added), it's just a little visual help. My take was, the better type checking for string etc. is worth it, just the visual hint isn't. But since more of you see it differently, let's go with your take.

Reply:
I completely agree. + But I think it would be useful to use this visual clue that PHP natively supports for places where there is no breakage change. 

I also find this discussion useful for improving the code. As I said, it's not critical for me, I can live without `mixed` 😅  I leave the final decision to you 🤝

Reply:
I'd say we can certainly do it for the helpers as it's not a breaking change here (functions cannot be extended).

In the future, we should always review if type hints are a breaking change (no matter if `string`, `mixed` or anything else). And that depends on whether the function/method can be extended by users or plugins. With the cache drivers it was the worst case, most of the time it's pretty much unlikely.